File: michangsheng/ItemSystem/items/loaders/items.py
import json
from pathlib import Path
from typing import Dict, List, Iterator
from pydantic import BaseModel, Field
from items.constants import TuJianType, WuPingType


class Item(BaseModel):
    """物品定义"""

    # 基础属性
    id: int = Field(description="ID")
    name: str = Field(description="物品名称(中文)")
    ItemIcon: int = Field(description="物品 Icon 序号")
    maxNum: int = Field(description="可持有的最大个数")
    type: WuPingType = Field(description="物品类型")
    quality: int = Field(description="品质")
    price: int = Field(description="基础售价")
    desc: str = Field(description="短描述")
    desc2: str = Field(description="长描述")
    CanSale: int = Field(description="是否能售卖, 0 表示可以")
    CanUse: int = Field(description="玩家是否能使用")
    NPCCanUse: int = Field(description="NPC 是否能使用")
    seid: List[int] = Field(description="作用效果列表")
    ItemFlag: List[int]
    Affix: List[int] = Field(description="词缀(id列表)")
    tu_jian_type: TuJianType = Field(description="图鉴类型, 用于解锁图鉴", alias="TuJianType")
    ShopType: int = Field(description="商品类型")

    FaBaoType: str = Field(description="未知用途")
    WuWeiType: int = Field(description="五维类型(炼器相关)")
    ShuXingType: int = Field(description="属性类型(炼器相关)")
    typePinJie: int = Field(description="")
    StuTime: int = Field(description="学习消耗时间")

    vagueType: int = Field(description="未知")
    DanDu: int = Field(description="丹毒")
    yaoZhi1: int = Field(description="作为药引的药质")
    yaoZhi2: int = Field(description="作为主药的药质")
    yaoZhi3: int = Field(description="作为副药的药质")
    wuDao: List = Field(description="等价的悟道经验")


class Items:
    _filename = "d_items.py.datas.json"

    def __init__(self):
        self._initialed = False
        self._items: Dict[int, Item] = {}

    def get_by_id(self, item_id: int) -> Item:
        self._load()
        return self._items[item_id]

    def filter_by_type(self, type: WuPingType) -> Iterator[Item]:
        self._load()
        for _, item in self._items.items():
            if item.type == type:
                yield item

    def _load(self):
        if self._initialed:
            return

        path = Path(__file__).parent.parent / "assets" / self._filename
        data = json.loads(path.read_text())
        for _, raw in data.items():
            item = Item(**raw)
            self._items[item.id] = item
        self._initialed = True


items = Items()

File: michangsheng/FriendlyLianDan/lib/loaders/items.cs
using System;
using System.Collections.Generic;
using UnityEngine;


namespace ItemSystem {
    [Serializable]
    public class Item {
        public int id { get; set; }
        public string name { get; set; }
        public int ItemIcon { get; set; }
        public int maxNum { get; set; }
        public WuPingType type { get; set; }
        public int quality { get; set; }
        public int price { get; set; }
        public string desc { get; set; }
        public string desc2 { get; set; }
        public int CanSale { get; set; }
        public int CanUse { get; set; }
        public int NPCCanUse { get; set; }
        public List<int> seid {get; set; }
        public List<int> ItemFlag {get; set; }
        public List<int> Affix {get; set; }
        public TuJianType TuJianType {get; set; }
        public int ShopType {get; set; }
        public string FaBaoType {get; set; }
        public string WuWeiType {get; set; }
        public string ShuXingType {get; set; }
        public string typePinJie {get; set; }
        public string StuTime {get; set; }
        public string vagueType {get; set; }
        public string DanDu {get; set; }
        public string yaoZhi1 {get; set; }
        public string yaoZhi2 {get; set; }
        public string yaoZhi3 {get; set; }
        public string wuDao {get; set; }

        public override string ToString() {
            return $"物品<{id}: {name}>";
        }
    }

    public class Items {
        private static Items instance = null;
        private static readonly object _lock = new object();
        private static string path = "Effect/json/d_items.py.datas";

        public IDictionary<int, Item> items;

        private Items () {
            items = new Dictionary<int, Item>{};
            string data = ModResources.LoadText(path);
            Dictionary<string, Item> loaded = JsonUtility.FromJson<Dictionary<string, Item>>(data);
            foreach (KeyValuePair<string, Item> kv in loaded) {
                items[kv.Value.id] = kv.Value;
            }
        }

        public static Items Instance {
            get {
                if (instance == null) {
                    lock(_lock) {
                        if (instance == null) {
                            instance = new Items();
                        }
                    }
                }
                return instance;
            }
        }
    }
}